src/bson.rs

```diff
@@ -556,12 +517,20 @@ impl Bson {
                     code: code.to_owned(),
                     scope: scope.to_owned(),
                 });
-            } else if let (Ok(t), Ok(i)) = (values.get_i32("t"), values.get_i32("i")) {
-                let timestamp = ((t as i64) << 32) + (i as i64);
-                return Bson::TimeStamp(timestamp);
-            } else if let (Ok(t), Ok(i)) = (values.get_i64("t"), values.get_i64("i")) {
-                let timestamp = (t << 32) + i;
-                return Bson::TimeStamp(timestamp);
+            } else if let (Ok(time), Ok(increment)) = (values.get_i32("t"), values.get_i32("i")) {
+                if time >= 0 && increment >= 0 {
+                    return Bson::TimeStamp(TimeStamp {
+                        time: time as u32,
+                        increment: increment as u32,
+                    });
+                }
+            } else if let (Ok(time), Ok(increment)) = (values.get_i64("t"), values.get_i64("i")) {
+                if time >= 0 && increment >= 0 {
+                    return Bson::TimeStamp(TimeStamp {
+                        time: time as u32,
```

Comment:
will this panic if either are too big?

Reply:
the MAX values are inclusive right? So we probably need `<=`.

Reply:
Updated